versioned_hdf5/subchunk_map.py

```diff
@@ -538,6 +903,254 @@ def _maybe_array_idx_to_slice(idx: Any):  # -> NDArray[np_hsize_t] | slice:
+        TransferType.setitem
+            The src array is the value parameter of __setitem__
+            The dst array is an individual slab
+            dst_slab_offsets must have as many rows as chunk_idxidx
+        TransferType.slab_to_slab
+            Both src and dst arrays are slabs
+
+    mappers:
+        List of IndexChunkMapper objects, one per axis of the index, as returned by
+        index_chunk_mappers()
+    chunk_idxidx:
+        2D array of shape (nchunks, ndim) where each row represents a chunk to be
+        transferred and point contains the index of IndexChunkMapper.chunk_indices for
+        the axis matching the column. In other words,
+
+            mappers[j].chunk_indices[chunk_idxidx[i, j]] * mappers[j].chunk_size
+
+        is the address along axis j of the top-left corner in the virtual array
+        of the i-th chunk to be transferred.
+
+    src_slab_offsets:
+        1D array of shape (nchunks, ) with the offset of the slab along axis 0, to be
+        added to addresses relative to the start of the chunk in the src slab.
+        Ignored when transfer_type==TransferType.setitem.
+    dst_slab_offsets:
+        1D array of shape (nchunks, ) with the offset of the slab along axis 0, to be
+        added to addresses relative to the start of the chunk in the dst slab.
+        Ignored when transfer_type==TransferType.getitem.
+
+    Returns
+    -------
+    3D array of indices of shape (nslices, 5, ndim), ready to be passed to
+    read_many_slices().
+
+    There is one row per slice.
+    In absence of fancy indexing, this is the same as the number of chunks.
+
+    Axis 1 contains exactly 5 columns:
+
+    - src_start
+    - dst_start
+    - count
+    - src_stride
+    - dst_stride
+
+    See slicetools.pyx::read_many_slices() for documentation on these parameters.
+    See subchunk_map.pxd::ReadManySlicesNDColumn for reference on the column indices.
+
+    The final column represents the axis. Note how axis must be last, and thus
+    internally C-contiguous, in order to allow read_many_slices() to use the index as-is
+    without any further processing.
+    """
+    # Hack around lack of cython enums in pure python mode
+    transfer_type_int: cython.int = transfer_type.value
+    TRANSFER_TYPE_GETITEM: cython.int = TransferType.getitem.value
+    TRANSFER_TYPE_SETITEM: cython.int = TransferType.setitem.value
+
+    nchunks = chunk_idxidx.shape[0]
+    ndim = chunk_idxidx.shape[1]
+
+    # Store the slices in a C array (in the stack) of views (in the heap).
+    # In order to keep the reference counter of the views, which are Python objects,
+    # above zero for the duration of the function, we also need to reference them from a
+    # Python object.
+    slices_refs = []
+    # Can't define module-level constants in pure-python mode
+    # https://github.com/numpy/numpy/blob/b0a52b76c2bebe5c237722589d13bf02affa9c43/numpy/core/include/numpy/ndarraytypes.h#L30
+    # NPY_MAXDIMS = 32
```

Comment:
?!

Reply:
defining an array in the stack below (`slices: hsize_t[:, :][32]`) requires knowing the exact size of the array at compile time. in C/C++ this is done through `#define NPY_MAXDIMS 32`. The Cython equivalent is deprecated: https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#deprecated-def-if

None of the suggested alternatives work in pure-python mode.

Reply:
https://github.com/cython/cython/issues/6505